lib/elixir/lib/module/types/descr.ex

```diff
@@ -495,4 +525,539 @@ defmodule Module.Types.Descr do
+        if tag == :closed and not neg_opt, do: throw(:no_intersection)
+      end
+
+      for {key, {is_opt, _type}} when not is_map_key(neg_fields, key) <- fields,
+          # key is required, and the negative map is closed: empty intersection
+          not is_opt and neg_tag == :closed do
+        throw(:no_intersection)
+      end
+
+      [{neg_tag, neg_fields} | filter_empty_negations({tag, fields}, negs)]
+    catch
+      :no_intersection -> filter_empty_negations({tag, fields}, negs)
+    end
+  end
+
+  defp map_line_to_quoted({positive_map, negative_maps}) do
+    case negative_maps do
+      [] ->
+        map_literal_to_quoted(positive_map)
+
+      _ ->
+        negative_maps
+        |> Enum.map(&map_literal_to_quoted/1)
+        |> Enum.reduce(&{:or, [], [&2, &1]})
+        |> List.wrap()
+        |> Kernel.then(&{:and, [], [map_literal_to_quoted(positive_map), {:not, [], [&1]}]})
+    end
+  end
+
+  def map_literal_to_quoted({tag, fields}) do
+    case tag do
+      :closed -> {:%{}, [], fields_to_quoted(tag, fields)}
+      :open -> {:%{}, [], [{:..., [], nil} | fields_to_quoted(tag, fields)]}
+    end
+  end
+
+  defp fields_to_quoted(tag, map) do
+    for {key, {optional_field?, type}} <- map,
+        not (tag == :open and optional_field? and term?(type)) do
+      cond do
+        optional_field? and empty?(type) -> {literal(key), {:not_set, [], []}}
+        optional_field? -> {literal(key), {:if_set, [], [to_quoted(type)]}}
+        true -> {literal(key), to_quoted(type)}
+      end
+    end
+  end
+
+  ## Pairs
+  #
+  # To simplify disjunctive normal forms of e.g., map types, it is useful to
+  # convert them into disjunctive normal forms of pairs of types, and define
+  # normalization algorithms on pairs.
+
+  # Takes a DNF of pairs and simplifies it into a equivalent single list (union)
+  # of type pairs. The `term` argument should be either `@term_or_unset` (for
+  # map value types) or `@term` in general.
+  # Remark: all lines of a pair dnf are naturally disjoint, because choosing a
+  # different edge means intersection with a literal or its negation.
+
+  # A line is a list of pairs `{positive, negative}` where `positive` is a list of
+  # literals and `negative` is a list of negated literals. Positive pairs can
+  # all be intersected component-wise. Negative ones are eliminated iteratively.
+
+  # Eliminates negations from `{t, s} and not negative` where `negative` is a
+  # union of pairs disjoint on their first component.
+  # Formula:
+  #   {t, s} and not (union<i=1..n> {t_i, s_i})
+  #       = union<i=1..n> {t and t_i, s and not s_i}
+  #            or {t and not (union{i=1..n} t_i), s}
+  # This eliminates all top-level negations and produces a union of pairs that
+  # are disjoint on their first component.
+  defp eliminate_negations(negative, t, s) do
+    {pair_union, union_of_t_i} =
+      Enum.reduce(
+        negative,
+        {[], none()},
```

Comment:
We should pass `t` instead of `none()` and already compute the difference as we move forward instead of `union_of_t_i`.